File: frontend/src/types/trading.ts
export interface TradingSession {
  id: string;
  user_id: string;
  strategy_id: string;
  strategy_name: string | null;
  strategy_version: number;
  mode: "paper" | "live";
  status: "stopped" | "running" | "paused" | "error";
  initial_capital: number;
  current_capital: number | null;
  parameters: Record<string, any>;
  instruments: string[];
  timeframe: string;
  error_message: string | null;
  started_at: string | null;
  stopped_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface TradingSessionListItem {
  id: string;
  strategy_id: string;
  strategy_name: string | null;
  strategy_version: number;
  mode: "paper" | "live";
  status: "stopped" | "running" | "paused" | "error";
  initial_capital: number;
  current_capital: number | null;
  instruments: string[];
  timeframe: string;
  started_at: string | null;
  created_at: string;
}

export interface CreateSessionRequest {
  strategy_id: string;
  mode: "paper" | "live";
  initial_capital?: number;
  parameters?: Record<string, any>;
  instruments?: string[];
  timeframe?: string;
}

export interface TradingSnapshot {
  session_id: string;
  status: string;
  portfolio_value: number;
  cash: number;
  total_pnl: number;
  positions: TradingPosition[];
  open_orders: number;
  total_trades: number;
  total_charges: number;
  prices: Record<string, number>;
}

export interface TradingPosition {
  symbol: string;
  exchange: string;
  side: string;
  quantity: number;
  avg_price: number;
  current_price: number;
  unrealized_pnl: number;
  pnl_percent: number;
  sl_price: number | null;
  tp_price: number | null;
  sl_order_id: string | null;
  tp_order_id: string | null;
}

export interface TradingOrder {
  id: string;
  tradingsymbol: string;
  exchange: string;
  transaction_type: string;
  order_type: string;
  product: string;
  quantity: number;
  price: number | null;
  filled_quantity: number;
  average_price: number | null;
  status: string;
  mode: string;
  placed_at: string;
  filled_at: string | null;
}

export interface TradingTrade {
  id: string;
  tradingsymbol: string;
  exchange: string;
  side: string;
  quantity: number;
  entry_price: number;
  exit_price: number | null;
  pnl: number | null;
  pnl_percent: number | null;
  charges: number;
  net_pnl: number | null;
  mode: string;
  entry_at: string;
  exit_at: string | null;
}

export interface SessionRunListItem {
  id: string;
  trading_session_id: string;
  run_number: number;
  status: "running" | "completed" | "error";
  initial_capital: number;
  final_capital: number | null;
  total_return: number | null;
  total_trades: number | null;
  win_rate: number | null;
  profit_factor: number | null;
  max_drawdown: number | null;
  started_at: string;
  stopped_at: string | null;
}

export interface SessionRun extends SessionRunListItem {
  cagr: number | null;
  sharpe_ratio: number | null;
  sortino_ratio: number | null;
  avg_trade_pnl: number | null;
  equity_curve: { timestamp: string; equity: number }[] | null;
  drawdown_curve: { timestamp: string; drawdown_percent: number }[] | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/trading.py
import uuid
from datetime import datetime
from pydantic import BaseModel, Field


class TradingSessionCreate(BaseModel):
    strategy_id: uuid.UUID
    mode: str = Field(..., pattern=r"^(paper|live)$")
    initial_capital: float = Field(default=100000.00, gt=0)
    parameters: dict = Field(default_factory=dict)
    instruments: list[str] = Field(default_factory=list)
    timeframe: str = Field(default="1d", pattern=r"^(1m|5m|15m|30m|1h|1d)$")


class TradingSessionResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    strategy_id: uuid.UUID
    strategy_name: str | None = None
    strategy_version: int
    mode: str
    status: str
    initial_capital: float
    current_capital: float | None
    parameters: dict
    instruments: list
    timeframe: str
    error_message: str | None
    started_at: datetime | None
    stopped_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TradingSessionListResponse(BaseModel):
    id: uuid.UUID
    strategy_id: uuid.UUID
    strategy_name: str | None = None
    strategy_version: int
    mode: str
    status: str
    initial_capital: float
    current_capital: float | None
    instruments: list
    timeframe: str
    started_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class OrderResponse(BaseModel):
    id: uuid.UUID
    tradingsymbol: str
    exchange: str
    transaction_type: str
    order_type: str
    product: str
    quantity: int
    price: float | None
    filled_quantity: int
    average_price: float | None
    status: str
    mode: str
    placed_at: datetime
    filled_at: datetime | None

    model_config = {"from_attributes": True}


class PositionResponse(BaseModel):
    id: uuid.UUID
    tradingsymbol: str
    exchange: str
    side: str
    quantity: int
    average_entry_price: float
    current_price: float | None
    unrealized_pnl: float | None
    mode: str

    model_config = {"from_attributes": True}


class TradeResponse(BaseModel):
    id: uuid.UUID
    tradingsymbol: str
    exchange: str
    side: str
    quantity: int
    entry_price: float
    exit_price: float | None
    pnl: float | None
    pnl_percent: float | None
    charges: float
    net_pnl: float | None
    mode: str
    entry_at: datetime
    exit_at: datetime | None

    model_config = {"from_attributes": True}
